pkg/server/home_workspaces.go

```diff
@@ -327,145 +181,182 @@ func (h *homeWorkspaceHandler) ServeHTTP(rw http.ResponseWriter, req *http.Reque
 
-			found, err := h.searchForHomeWorkspaceRBACResourcesInLocalInformers(homeLogicalClusterName)
-			if err != nil {
-				responsewriters.InternalError(rw, req, err)
-				return
-			}
-			if found && homeClusterWorkspace.Status.Phase == tenancyv1alpha1.ClusterWorkspacePhaseReady {
-				// We don't need to check any permission before returning the home workspace definition since,
-				// once it has been created, a home workspace is owned by the user.
-				homeWorkspace := &tenancyv1beta1.Workspace{}
-				projection.ProjectClusterWorkspaceToWorkspace(homeClusterWorkspace, homeWorkspace)
-				responsewriters.WriteObjectNegotiated(homeWorkspaceCodecs, negotiation.DefaultEndpointRestrictions, tenancyv1beta1.SchemeGroupVersion, rw, req, http.StatusOK, homeWorkspace)
+		if this.Status.Phase != tenancyv1alpha1.ClusterWorkspacePhaseReady {
+			if time.Since(this.CreationTimestamp.Time) > h.creationTimeout {
+				responsewriters.InternalError(rw, req, fmt.Errorf("home workspace creation timeout"))
 				return
 			}
 
-			// fall through because either not ready or RBAC objects missing
+			rw.Header().Set("Retry-After", fmt.Sprintf("%d", h.creationDelaySeconds))
+			http.Error(rw, "Creating the home workspace", http.StatusTooManyRequests)
+			return
 		}
 
-		// fall-through and let it be created
-		lcluster.Name = homeLogicalClusterName
-		workspaceType = HomeClusterWorkspaceType
-	} else {
-		if !lcluster.Name.HasPrefix(h.homePrefix) {
-			// Not a home workspaces request
-			h.apiHandler.ServeHTTP(rw, req)
-			return
+		// nothing to check permission-wise. If there is a very improbable hash conflict, the will get a 403
+		// through normal authorization. We don't give access to legacy home workspaces to new users here.
+
+		homeWorkspace := &tenancyv1beta1.Workspace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: this.Name,
+			},
+			Spec: tenancyv1beta1.WorkspaceSpec{
+				Type: this.Spec.Type,
+			},
+			Status: tenancyv1beta1.WorkspaceStatus{
+				URL:          this.Status.URL,
+				Phase:        this.Status.Phase,
+				Conditions:   this.Status.Conditions,
+				Initializers: this.Status.Initializers,
+			},
 		}
+		responsewriters.WriteObjectNegotiated(homeWorkspaceCodecs, negotiation.DefaultEndpointRestrictions, tenancyv1beta1.SchemeGroupVersion, rw, req, http.StatusOK, homeWorkspace)
+		return
+	}
 
-		var needsAutomaticCreation bool
-		needsAutomaticCreation, workspaceType = h.needsAutomaticCreation(lcluster.Name)
-		logger.V(4).Info("not a ~ request", "needsAutoCreate", needsAutomaticCreation, "verb", requestInfo.Verb, "url", req.URL)
-		if !needsAutomaticCreation {
-			h.apiHandler.ServeHTTP(rw, req)
+	// here we have not found a legacy home workspace. We have to check whether a new-style workspace exists.
+
+	homeClusterName := homeClusterName(effectiveUser.GetName())
+	var this *tenancyv1alpha1.ThisWorkspace
+	objs, err := h.thisWorkspaceIndexer.ByIndex(indexers.ByLogicalCluster, homeClusterName.String())
```

Comment:
how a user can end up on this particular shard?

Reply:
Good question. I think we need a counterpart in the front-proxy. Without front-proxy, the shard that gets the request is the chosen one. But in the multi-shard case, we need another entity doing the scheduling: the front-proxy.

And we must be careful with races. The front-proxy is eventually consistent. We have to make sure that it does not schedule the user again to another shard, such that it would have multiple home workspaces.